File: Problems/No.92_Reverse_Linked_List_II/reverse_linked_list_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseBetween(self, head, left, right):
        """
        :type head: ListNode
        :type left: int
        :type right: int
        :rtype: ListNode
        """

        # parameters
        curr, prev = head, None
        before_left, reverse_end, reverse_next = None, None, None

        # start computing
        if (left == right):
            return head
        
        # find the left node
        for i in range(1, left):
            prev = curr
            curr = curr.next
        
        # start reversing
        before_left = prev
        reverse_end = curr
        reverse_next = curr.next
        for i in range(right-left+1):
            curr.next = prev
            prev = curr
            curr = reverse_next
            if reverse_next:
                reverse_next = reverse_next.next
        
        # reconnection
        if before_left:
            before_left.next = prev
        else:
            head = prev
        reverse_end.next = curr
    
        return head

File: Problems/No.92_Reverse_Linked_List_II/reverse_linked_list_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        
        // parameters
        ListNode *curr = head, *prev = NULL;
        ListNode *before_left, *reverse_end, *reverse_next;

        // start computing
        if (left == right) return head;

        // find the left node
        for (int i=1; i<left; i++) {

            prev = curr;
            curr = curr->next;

        }

        // start reversing
        before_left = prev;
        reverse_end = curr;
        reverse_next = curr->next;
        for (int i=0; i<(right-left+1); i++) {

            curr->next = prev;
            prev = curr;
            curr = reverse_next;
            if (reverse_next != NULL) reverse_next = reverse_next->next;

        }

        // reconnection
        if (before_left != NULL) before_left->next = prev;
        else head = prev;
        reverse_end->next = curr;

        return head;

    }
};
